File: app/src/main/java/de/nicidienase/chaosflix/EventDetailsDescriptionPresenter.java
package de.nicidienase.chaosflix;

import android.content.Context;
import android.support.v17.leanback.widget.AbstractDetailsDescriptionPresenter;
import android.support.v17.leanback.widget.Presenter;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import butterknife.ButterKnife;
import de.nicidienase.chaosflix.entities.recording.Event;
import de.nicidienase.chaosflix.entities.streaming.Room;

/**
 * Created by felix on 18.03.17.
 */

public class EventDetailsDescriptionPresenter extends Presenter {

	private final Context mContext;

	public EventDetailsDescriptionPresenter(Context context){
		mContext = context;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent) {
		View view = LayoutInflater.from(mContext).inflate(R.layout.detail_view_content, null);
        return new ViewHolder(view);
	}

	@Override
	public void onBindViewHolder(ViewHolder vh, Object item) {
		TextView titleText = ButterKnife.findById(vh.view,R.id.title_text);
		TextView speakersText = ButterKnife.findById(vh.view,R.id.speakers_text);
		TextView subtitleText = ButterKnife.findById(vh.view,R.id.subtitle_text);
		TextView descriptionText = ButterKnife.findById(vh.view,R.id.description_text);
		if(item instanceof Event){
			Event event = (Event) item;

			titleText.setText(event.getTitle());
			subtitleText.setText(event.getSubtitle());
			String speaker = TextUtils.join(", ", event.getPersons());
			speakersText.setText(speaker);
			StringBuilder sb = new StringBuilder();
			sb.append(event.getDescription())
					.append("\n")
					.append("\nreleased at: ").append(event.getReleaseDate())
					.append("\nTags: ").append(android.text.TextUtils.join(", ",event.getTags()));
			descriptionText.setText(sb.toString());
		} else if(item instanceof Room){
			Room room = (Room) item;
			titleText.setText(room.getDisplay());
			subtitleText.setText(room.getShedulename());
		}
	}

	@Override
	public void onUnbindViewHolder(ViewHolder viewHolder) {

	}
}

File: leanback/src/main/java/de/nicidienase/chaosflix/leanback/detail/EventDetailsDescriptionPresenter.kt
package de.nicidienase.chaosflix.leanback.detail

import android.content.Context
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.leanback.widget.Presenter
import de.nicidienase.chaosflix.common.mediadata.entities.recording.persistence.Event
import de.nicidienase.chaosflix.common.mediadata.entities.streaming.Room
import de.nicidienase.chaosflix.leanback.databinding.DetailViewBinding

class EventDetailsDescriptionPresenter(private val context: Context) : Presenter() {

    override fun onCreateViewHolder(parent: ViewGroup): ViewHolder {
        val binding = DetailViewBinding.inflate(LayoutInflater.from(context))
        return DescriptionViewHolder(binding.root, binding)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, item: Any) {
        if (viewHolder !is DescriptionViewHolder) { context
            throw IllegalStateException("Wrong ViewHolder")
        }
        viewHolder.binding.item = when (item) {
            is Event -> {
                val sb = StringBuilder()
                val speaker = TextUtils.join(", ", item.persons ?: emptyArray())
                sb.append(item.description)
                        .append("\n")
                        .append("\nreleased at: ").append(item.releaseDate)
                        .append("\nTags: ").append(TextUtils.join(", ", item.tags!!))
                DetailDataHolder(item.title,
                        item.subtitle,
                        speaker,
                        sb.toString())
            }
            is Room -> {

                val currentTalk = item.talks?.get("current")
                val nextTalk = item.talks?.get("next")
                DetailDataHolder(title = item.display,
                        subtitle = item.schedulename,
                        speakers = currentTalk?.description ?: "",
                        description = "Next Talk: ${nextTalk?.description ?: "no talk scheduled"}")
            }
            else -> {
                Log.e(TAG, "Item is neither PersistentEvent nor Room, this should not be happening")
                DetailDataHolder("", "", "", "")
            }
        }
    }

    inner class DetailDataHolder internal constructor(val title: String, val subtitle: String?, val speakers: String, val description: String) {

        internal constructor(event: Event) : this(
                event.title,
                event.subtitle,
                TextUtils.join(", ", event.persons!!),
                StringBuilder().append(event.description)
                        .append("\n")
                        .append("\nreleased at: ").append(event.releaseDate)
                        .append("\nTags: ")
                        .append(android.text.TextUtils.join(", ", event.tags!!))
                        .toString())

        internal constructor(room: Room) : this(
                room.display,
                room.schedulename,
                "",
                "")
    }

    inner class DescriptionViewHolder(view: View, val binding: DetailViewBinding) : ViewHolder(view)

    override fun onUnbindViewHolder(vh: ViewHolder) {}

    companion object {
        private val TAG = EventDetailsDescriptionPresenter::class.java.simpleName
    }
}
